### DISCUSSION TITLE
cpp/object_oriented/initialization_and_destruction_order

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#53] [Announcements]
# cpp/object_oriented/initialization_and_destruction_order

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/object_oriented/initialization_and_destruction_order.html

**Comment by huangjiawei128** [comment]
补充一点：
初始化顺序：成员对象的初始化顺序按照它们在类中声明的顺序进行，而不是初始化列表中的顺序。这是为了保证对象构造的一致性和预见性。

初始化优先级：如果成员对象在声明时进行了初始化，并且在构造函数的初始化列表中也进行了初始化，构造函数的初始化列表中的初始化会覆盖声明时的初始化。

**Reply by huangjiawei128** [reply]
调用 ptr->~T(); 会调用 T 类型对象的析构函数。如果 T 类型对象包含成员对象属性，那么这些成员对象的析构函数也会被正确调用。这是因为当你调用一个对象的析构函数时，该对象的所有成员变量的析构函数也会被递归调用。
